pkg/cloudprovider/aws/fleet/packing/instancetype.go

```diff
@@ -0,0 +1,159 @@
+			},
+			pricePerHour: 1.536,
+		},
+		// {
+		// 	name: "m5.4xlarge	",
+		// 	totalCapacity: v1.ResourceList{
+		// 		v1.ResourceCPU:    resource.MustParse("16000m"),
+		// 		v1.ResourceMemory: resource.MustParse("64Gi"),
+		// 	},
+		// 	availableCapacity: v1.ResourceList{
+		// 		v1.ResourceCPU:    resource.MustParse("16000m"),
+		// 		v1.ResourceMemory: resource.MustParse("64Gi"),
+		// 	},
+		// 	pricePerHour: 0.768,
+		// },
+		{
+			name: "m5.2xlarge",
+			totalCapacity: v1.ResourceList{
+				v1.ResourceCPU:    resource.MustParse("8000m"),
+				v1.ResourceMemory: resource.MustParse("32Gi"),
+			},
+			availableCapacity: v1.ResourceList{
+				v1.ResourceCPU:    resource.MustParse("8000m"),
+				v1.ResourceMemory: resource.MustParse("32Gi"),
+			},
+			pricePerHour: 0.384,
+		},
+		{
+			name: "m5.xlarge",
+			totalCapacity: v1.ResourceList{
+				v1.ResourceCPU:    resource.MustParse("4000m"),
+				v1.ResourceMemory: resource.MustParse("16Gi"),
+			},
+			availableCapacity: v1.ResourceList{
+				v1.ResourceCPU:    resource.MustParse("4000m"),
+				v1.ResourceMemory: resource.MustParse("16Gi"),
+			},
+			pricePerHour: 0.192,
+		},
+		{
+			name: "m5.large",
+			totalCapacity: v1.ResourceList{
+				v1.ResourceCPU:    resource.MustParse("2"),
+				v1.ResourceMemory: resource.MustParse("8Gi"),
+			},
+			availableCapacity: v1.ResourceList{
+				v1.ResourceCPU:    resource.MustParse("2"),
+				v1.ResourceMemory: resource.MustParse("8Gi"),
+			},
+			pricePerHour: 0.096,
+		},
+	}
+)
+
+type instanceType struct {
+	name string
+	sync.RWMutex
+	availableCapacity v1.ResourceList
+	totalCapacity     v1.ResourceList
+	pricePerHour      float64
+}
+
+func (p *PodPacker) getInstanceTypes(filter string) []*instanceType {
+	instanceTypeOptions := instanceTypesAvailable
+	// sort by increasing capacity of the instance
+	sort.Sort(byCapacity{instanceTypeOptions})
+	return instanceTypeOptions
+}
+
+func (r *instanceType) isAllocatable(cpu, memory resource.Quantity) bool {
+	r.Lock()
+	defer r.Unlock()
+	return (r.availableCapacity.Cpu().Cmp(cpu) >= 0) &&
+		(r.availableCapacity.Memory().Cmp(memory) >= 0)
+}
+
+func (r *instanceType) reserveCapacity(cpu, memory resource.Quantity) error {
+	r.Lock()
+	defer r.Unlock()
+	if r.availableCapacity.Cpu().IsZero() {
+		return fmt.Errorf("insufficient CPU capacity")
+	}
+	if r.availableCapacity.Memory().IsZero() {
+		return fmt.Errorf("insufficient Memory capacity")
+	}
+	r.availableCapacity.Cpu().Sub(cpu)
+	r.availableCapacity.Memory().Sub(memory)
+	return nil
+}
+
+func filterInstancesBasedOnCost(instanceTypeOptions []*instanceType) []*instanceType {
+	sort.Sort(byCost{instanceTypeOptions})
+	return instanceTypeOptions
+}
+
+type instanceTypes []*instanceType
+
+func (c instanceTypes) Len() int {
+	return len(c)
+}
+
+func (c instanceTypes) Swap(i, j int) {
+	c[i], c[j] = c[j], c[i]
+}
+
+type byCapacity struct{ instanceTypes }
+
+func (c byCapacity) Less(i, j int) bool {
```

Comment:
It looks like this function slightly favors/focuses on CPU optimization over Memory optimization. (if CPU is equal, then use Memory, use CPU comparison otherwise)

Is this an opinion on limiting options or is there a technical reason for this?

Reply:
yes, CPU is preferred over memory when scoring based on resources requested, its based on the same logic how 2cpu and 16gb is a large (`r5.large`) instance type, however, 4cpu and 8gb is an xlarge (`c5.xlarge`).

Reply:
So would it be fair to say that our heuristic here is to pick the smallest instance type of a family? Do the smallest instance sizes of a family have the best cost efficiency?

Reply:
Yes, start with the smallest instance and see which instance type gives the best packing, i.e. the type that can pack the most pods passed in the given `Pack()` call. If there are multiple such instance types, pass them as an option to create fleet request and fleet API takes care of selecting the cheapest instance type from the list provided.